### PR TITLE
Implementation of feedback and bug reports API

### PR DESCRIPTION
**Pull request by johan1505**
1. Added a design document for the Feedback Portal Micro-service. Implemented the routes for the feedback and bug reports APIs. Also, added tests for the routes.
2.  Routes only contain the logic for getting data from the request and sending a response back to the client. Logic related to CRUD operations performed on the MongoDB collections is encapsulated in separate modules 
3. Package added: Faker (to fake parameters in tests)
4. Code coverage is not 100% in the route files, but it is pretty close to that.

### CODE REVIEW

File: src/modules/bug-reports.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+// We'll assume the limit is always 10 for now
+// If the page number exceeds the number of available pages, 0 reports are returned
+
+const numberOfDocumentsPerPage = 10;
+export const getReports = (page: number): Promise<BugReport[]> => {
+    return (
+        Collections.BugReport()
+            .find()
+            .sort({ date: -1 })
```
Comment (DevelopmentByDavid) [inline src/modules/bug-reports.ts:34]: Make this a query option

File: src/routes/bug-reports/bug-reports.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    }
+});
+
+interface UserRequestBody {
+    user: User;
+}
+
+router.get('/get-reports/:submitterId', async (req: Request, res: Response) => {
+    try {
+        const { submitterId } = req.params;
```
Comment (DevelopmentByDavid) [inline src/routes/bug-reports/bug-reports.ts:76]:
```js
const { submitterId } = req.params as { submitterId: string };
```

File: src/routes/bug-reports/bug-reports.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        res.status(200).send({ reports: bugReports });
+    } catch (error) {
+        log.error(error);
+        res.statusMessage = 'Some error occurred. Please try again';
+        res.sendStatus(400);
+    }
+});
+
+interface UserRequestBody {
+    user: User;
```
Comment (DevelopmentByDavid) [inline src/routes/bug-reports/bug-reports.ts:71]:
```js
user?: User
```

File: src/routes/feedback-reports/feedback-reports.test.ts [hunk 4]
```diff
@@ -0,0 +1,696 @@
...
...eefwefv517248bfct71284v5tg246758ob234b1fwwevweefwefvvn4751b514nr89ptyew8vtyn12485vn4751b517248bfct71284v5tg246758ob234b1fwwevweefwefv514nr89ptyew8vtyn12485vn4751b517248bfct71284v5tg246758ob234b1fwwevweefwefv517248bfct71284v5tg246758ob234b1fwwevweefwefvw8vtyn12485vn4751b517248bfct71284v5tg246758ob234b514nr89ptyew8vtyn12485vn4751b514nr89ptyew8vtyn12485vn4751b517248bfct71284v5tg246758ob234b1fwwevweefwefv514nr89ptyew8vtyn12485vn4751b517248bfct71284v5tg246758ob234b1fwwevweefwefv517248bfct71284v5tg246758ob234b1fwwevweefwefv1fwwevweefw514nr89ptyew8vtyn12485vn4751b514nr89ptyew8vtyn12485vn4751b517248bfct71284v5tg246758ob234b1fwwevweefwefv514nr89ptyew8vtyn12485vn4751b517248bfct71284v5tg246758ob234b1fwwevweefwefv517248bfct71284v5tg246758ob234b1fwwevweefwefvefv514nr89ptyew8vtyn12485vn4751b517248bfct71284v5tg246758ob234b1fwwevweefwefv517248bfct71284v5tg246758ob234b1fwwevweefwefvfwwevweefwefv517248bfct71284v5tg246758ob234b1fwwevweefwefvefv514nr89ptyew8vtyn12485vn4751b517248bfct71284v5tg246758ob234b1fwwevweefwefv517248bfct71284v5tg246758ob234b1fwwevweefwefv234b1fwwevweefwefvefv514nr89ptyew8vtyn12485vn4751b517248bfct71284v5tg246758ob234b1fwwevweefwefv517248bfct71284v5tg246758ob234b1fwwevweefwefv',
```
[the remark this answers is not in the document]
Comment (DevelopmentByDavid) [reply src/routes/feedback-reports/feedback-reports.test.ts:129]: https://www.npmjs.com/package/faker

File: docker-compose.test.yaml [hunk 5]
```diff
@@ -1,21 +1,23 @@
...
+            - mongo
         #     - postgresql
         #     - rabbitmq
         #     - redis
     ###### uncomment the correponsing db's/services you need
-    # mongo:
-    #     image: mongo:4.2.8-bionic
+    mongo:
+        image: mongo:4.2.8-bionic
+        ports:
```
Comment (DevelopmentByDavid) [inline docker-compose.test.yaml:27]: Lets change the ports here in this file and the `docker-compose.test.yaml`.  If someone has a locally running mongo instance (as I do), then there is a conflict.

File: docker-compose.yaml [hunk 6]
```diff
@@ -7,14 +7,16 @@ services:
...
         #     - postgresql
         #     - rabbitmq
         #     - redis
     ###### uncomment the correponsing db's/services you need
-    # mongo:
-    #     image: mongo:4.2.8-bionic
+    mongo:
+        image: mongo:4.2.8-bionic
+        ports:
+            - 27017:27017
```
Comment (DevelopmentByDavid) [inline docker-compose.yaml:28]: refer to previous comment

File: src/routes/index.ts [hunk 7]
```diff
@@ -1,7 +1,10 @@
 import express from 'express';
+import feedbackReportsRoutes from './feedback-reports/feedback-reports';
+import bugReportsRoutes from './bug-reports/bug-reports';
```
Comment (DevelopmentByDavid) [inline src/routes/index.ts:3]: You should be putting an `index.ts` in each of those folders so you can just do `import xxxx from './xxxx-reports';`

File: src/routes/index.ts [hunk 8]
```diff
@@ -1,7 +1,10 @@
...
 
 const router = express.Router();
 
-router.get('/hello-world', (req, res) => res.send('Hello world!'));
+router.use('/feedback', feedbackReportsRoutes);
+router.use('/bugs', bugReportsRoutes);
```
Comment (DevelopmentByDavid) [inline src/routes/index.ts:8]: Make this `/api/feedback` and `/api/bugs` respectively

File: src/lib/interfaces.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    description: string;
+    submitterId: string;
+}
+
+export interface User {
+    // TODO: Change to mongoDb object ID if necessary in the future
+    _id: string;
+}
+
+export interface FeedbackReport extends Report {
```
Comment (DevelopmentByDavid) [inline src/lib/interfaces.ts:14]: Eslint is probably yelling at you here, just add a eslint-ignore here.

File: docker-compose.yaml (hunk 6 above)
Comment (johan1505) [reply docker-compose.yaml:28]: Basically you want docker.compose and docker.compose.test to run instances of mongoDB at different ports, correct?

**Comment by johan1505** [comment]
> For clarification, are all of the console.error's I see when running the tests the ones from using `src/lib/log.ts`?
> 
> Also, make sure to pull. I added two changes.

Yes, all of them use log.ts to display error messages

File: docker-compose.yaml (hunk 6 above)
Comment (DevelopmentByDavid) [reply docker-compose.yaml:28]: Yeah, basically.

File: .eslintignore [hunk 10]
```diff
@@ -1 +1,2 @@
-jest.config.js
\ No newline at end of file
+jest.config.js
+/src/lib/interfaces.ts
```
Comment (DevelopmentByDavid) [inline .eslintignore:2]: Oh, I see I just meant adding a `// eslint-disable-next-line @typescript-eslint/no-empty-interface` above the line of code I mentioned.  We don't want to ignore the whole file.

File: docker-compose.test.yml [hunk 11]
```diff
@@ -1,21 +1,31 @@
...
         #     - postgresql
         #     - rabbitmq
         #     - redis
     ###### uncomment the correponsing db's/services you need
-    # mongo:
-    #     image: mongo:4.2.8-bionic
+    mongo:
+        image: mongo:4.2.8-bionic
+        ports:
+            - 27018:27018
```
Comment (DevelopmentByDavid) [inline docker-compose.test.yml:28]: This is `host port:exposed port`.  Mongo should expose 27017 by default I believe (unless they just expose all ports on this container).  So all you would actually need to change is the host port.  Ex. `3005:27017` would map the exposed port 27017 on the mongo container to host port 3005.  I'm not sure that those ports currently would work.

File: .eslintignore (hunk 10 above)
Comment (johan1505) [reply .eslintignore:2]: Oh I see

**Pull request merged**